### PR TITLE
Some love for the rustbook (engine)

### PR DESCRIPTION
**Pull request by azerupi**
So I have tried to improve the rustbook engine:
- The sidebar now looks a lot more like gitbook (I thinks it cleaner)
- Added the Open Sans font, in my opinion more readable for prolonged periods of time
- Changed the style for code blocks a little

I encountered 1 problem. In `build.rs` I added this google font url (I commented out the non-relevant parts for clarity) 

``` rust
let rustdoc_args: &[String] = &[
    //"".to_string(),
    //preprocessed_path.display().to_string(),
    //format!("-o{}", out_path.display()),
    //format!("--html-before-content={}", prelude.display()),
    //format!("--html-after-content={}", postlude.display()),
    //format!("--markdown-playground-url=http://play.rust-lang.org"),
    //format!("--markdown-css={}", item.path_to_root.join("rust-book.css").display()),
    format!("--markdown-css=http://fonts.googleapis.com/css?family&#61;Open+Sans:400italic,700italic,400,700"),
    //"--markdown-no-toc".to_string(),
];
```

As you can see, I had to escape `=` with `&#61;` because the string would get truncated if I didn't. Is that normal behaviour? Is that for security measures? If it is, isn't it a little weak if you can circumvent it by escaped characters? I don't know the reason behind, but I thought it was at least worth mentioning :)

Take your time for this PR, I still want to add multiple improvements:
- Like gitbook, possibility to change font by user
- Put `css` and `js` in their respective files (not hardcoded in rust)
- button to hide sidebar
- ...

So I'm not in a hurry to get this merged ;) But if you think it's good enough to be merged, go ahead. I will make another PR when I have other improvements.

In the image below is a screen of the improvements

rustbook

### CODE REVIEW

**Comment by azerupi** [comment]
I want to add some icons.
At the moment I only need two. One to let the user modify the font and one to hide the sidebar. Is is ok if I add [FontAwesome](http://fortawesome.github.io/Font-Awesome/icons/) ?

They have a CDN so I don't need to include the icons in the rust repo, I can just link to their CDN (like I do for the Open Sans google font)

**Comment by steveklabnik** [comment]
it's important that docs render well when not-online, so I am tempted to say no to FontAwesome. There are some that already argue our existing fonts are too much...

**Comment by alexcrichton** [comment]
Nice improvements! A few thoughts:

* I agree with @steveklabnik, can we just stick to default fonts for now? We haven't had a great track record with custom fonts.
* Could you put up a copy online of the entire book to browse and compare?

**Comment by azerupi** [comment]
Oh that's a pity, it would have added a nice touch instead of menus made of words.

I don't see their point though, since everything is packed with the docs it doesn't matter much how much font's we use. The only argument I could understand is that it increases page load (but that's only relevant for the online version) and you can improve that by using CDN's + local fallback. 

Is there another argument against it?

@alexcrichton I used Open Sans from google fonts so it's not added to the repo. It does show up when you browse locally and have a an internet connection. But if it doesn't load it's no problem at all because you fall back on the old font ;)

> Could you put up a copy online of the entire book to browse and compare?

I think I can :) I will post a link when it's done

**Comment by azerupi** [comment]
@alexcrichton @steveklabnik here is the link: http://www.mathieudavid.org/test/rustbook/patch1/

I am working on a version to change font style at the moment I have included FontAwesome through CDN link I will add a local fallback to use words instead of icons when it doesn't load (for offline support). Once done I will put it online for you to review.
If then you still think that FontAwesome should not be included I will remove it and do only word menu :)

**Comment by azerupi** [comment]
I would need some help with separating the css and js from rust. At the moment it is hardcoded in a Rust string and not very maintainable. But I'm not sure what I have to modify. I guess I will have to modify the make file to copy over those static assets to the "production" folder, but I have no clue how?

Also it would be super convenient to add a make target like `make trpl-static` that would just copy the static files without rebuilding the rust code, for when you only modified the css or js.

**Comment by azerupi** [comment]
I have effectively isolated the css and js from the rust code. I put them in a static folder. Since `css.rs` wasn't used anymore I deleted it, was it used for something else?

It seems to behave correctly but I would feel a hell of a lot more confident if you guys could double check it ;)

**Comment by azerupi** [comment]
@steveklabnik @alexcrichton  any update on this? :)

File: src/rustbook/javascript.rs [hunk 1]
```diff
@@ -11,65 +11,7 @@
...
-        nav.appendChild(nextNode);
-      }
-      document.getElementById('page').appendChild(nav);
-      break;
-    }
-  }
-
-});
-</script>
+<<<<<<< HEAD
```
Comment (nhowell) [inline src/rustbook/javascript.rs]: This line looks like junk leftover from a merge.

File: src/rustbook/build.rs [hunk 2]
```diff
@@ -145,6 +145,8 @@ fn render(book: &Book, tgt: &Path) -> CliResult<()> {
             format!("--html-after-content={}", postlude.display()),
             format!("--markdown-playground-url=http://play.rust-lang.org"),
             format!("--markdown-css={}", item.path_to_root.join("rust-book.css").display()),
+            format!("--markdown-css={}",
+            "http://fonts.googleapis.com/css?family&#61;Open+Sans:400italic,700italic,400,700"),
```
Comment (nhowell) [inline src/rustbook/build.rs]: For style, this line needs to be indented a level, I believe.

File: src/llvm [hunk 3]
```diff
@@ -1 +1 @@
-Subproject commit 8cbcdf1b72e1b23679646f6faca265f76b20d379
+Subproject commit bff69076975642c64e76dbeaa53476bfa7212086
```
Comment (nhowell) [inline src/llvm]: This PR shouldn't be modifying this, right?

File: src/rust-installer [hunk 4]
```diff
@@ -1 +1 @@
-Subproject commit 8e4f8ea581502a2edc8177a040300e05ff7f91e3
+Subproject commit e54d4823d26cdb3f98e5a1b17e1c257cd329aa61
```
Comment (nhowell) [inline src/rust-installer]: And shouldn't be modifying this either, right?

File: src/llvm (hunk 3 above)
Comment (azerupi) [inline src/llvm]: I have no idea how this got modified, it probably got messed up when I merged the upstream branch to resolve the merge conflict.

File: src/rustbook/javascript.rs (hunk 1 above)
Comment (azerupi) [inline src/rustbook/javascript.rs]: Oh, yes my bad. I completely overlooked that. Correcting it right away :)

File: src/llvm (hunk 3 above)
Comment (nhowell) [inline src/llvm]: Okay. Yeah... that merge commit is pretty nasty looking. I wonder if a rebase and squash would take care of it (and then revert the changes to "src/llvm" and "src/rust-installer")

Comment (azerupi) [inline src/llvm]: I have never done that, could you walk me through the procedure?

Comment (nhowell) [inline src/llvm]: To rebase (I hope I have this correct):
````
$ git checkout doc-experiments
$ git fetch upstream
$ git rebase upstream/master
````
Let's see if that helps first. Squashing is a little more involved.

Comment (azerupi) [inline src/llvm]: Thank you ! It looks ok, I will check further and do a test build to see if everything is still good before I upload this time ;)

**Comment by azerupi** [comment]
@nhowell The rebase worked, I think it's all good now. Thanks for your help !

**Comment by steveklabnik** [comment]
I think this change is overall good, but still am opposed to Yet More Web Fonts. We get a lot of complaints about our existing usage, I'd rather be conservative for now.

Other than that, r=me

**Comment by azerupi** [comment]
@steveklabnik  Ok, I will remove the font then :) But I would like to explain why (in my opinion) it has only advantages. Could this maybe be discussed on the internal forum, to see what others think about it ?

- It's not downloaded locally, meaning that it's not yet another dependency
- If this font fails to load (for example offline), the next font on the list is used, that is the current font. So it's not like it would fail to render.
- Open Sans is one of the most widely used fonts from google CDN, there is a very very big chance that the user already has it in his browser cache, so he probably doesn't even need to reload it (= improved page load)

And I think it's easier on the eye, but that's a matter of personal taste ;)

**Comment by steveklabnik** [comment]
> I would like to explain why (in my opinion) it has only advantages.

A _lot_ of people complain about our existing web font usage. I'd prefer to keep that usage level at about the same until we see if maybe someday we remove it entirely. We really need some help from some designer people :(

I don't know about any changes to `make TRPL`, but I bet an LLVM upgrade happened, which would trigger a full rebuild until after you've done it once. `NO_REBUILD=1` sometimes helps with this, sometimes not...

**Comment by steveklabnik** [comment]
(Also, thanks so much for sticking with this!)

**Comment by azerupi** [comment]
> (Also, thanks so much for sticking with this!)

It is my pleasure, I'm glad I was able to help, even if it's not very significant. I hope I can do more in the future :)

> A lot of people complain about our existing web font usage. I'd prefer to keep that usage level at about the same

Ok fair enough :)

> We really need some help from some designer people

I am not going to pretend I am a designer, but I have played a lot with HTML 5, CSS and Javascript the last couple of years. I could be able to help out in some areas if there is need for this :) 
For example, [here is a little webapp](http://www.mathieudavid.org/webapps/sotawatchfilter/#/spotfilter) I made for my uncle last year. 

If you have something specific, I would be glad to see if I can help.


*PS: don't look at my website (homepage), I have been (re)designing my site since I put it online last year but haven't had enough time to complete it yet ;)*

**Comment by steveklabnik** [comment]
>  I could be able to help out in some areas if there is need for this :) 

That'd be great! There's lots of stuff that's not huge. For example, https://github.com/rust-lang/rust/issues?q=is%3Aopen+is%3Aissue+label%3AA-rustdoc has a huge list of things, many of which are web-dev style stuff, like https://github.com/rust-lang/rust/issues/26016 and https://github.com/rust-lang/rust/issues/24899 and https://github.com/rust-lang/rust/issues/23900

**Comment by azerupi** [comment]
Nice, I will see what I can do :)

**Comment by azerupi** [comment]
How stupid can I be.. I deleted the branch, luckily github has an awesome restore feature ! :P

**Comment by dotdash** [comment]
@bors r-

Seems that the delete/restore action made bors think this was approved again

**Comment by azerupi** [comment]
My apologies for the trouble.
I was working on another patch and wanted to clean up my fork a little bit.. 

*Note to myself, don't delete random branches when only half awake ;)*

**Comment by steveklabnik** [comment]
I think this was just a flaky test

**Comment by azerupi** [comment]
What is making the test fail?

**Comment by azerupi** [comment]
@steveklabnik what should I do?

**Comment by steveklabnik** [comment]
You didn't change any of the makefiles, so this must be a flaky test?

**Comment by azerupi** [comment]
Ok, glad to hear it's not my fault :)

**Comment by steveklabnik** [comment]
Okay, I have no idea here. @alexcrichton @brson do either of you know why
this would start failing?

**Comment by brson** [comment]
My guess is that [this command](https://github.com/rust-lang/rust/pull/26216/files#diff-d4615faf4ae9dfb7f326c10ebc4fae1fL200) to copy static assets from a relative directory is failing. Does this patch work with out-of-tree builds?

**Comment by brson** [comment]
Thanks for the rustbook love. It could be a swell tool.

**Comment by azerupi** [comment]
> My guess is that this command to copy static assets from a relative directory of cwd is failing.

What would cause it to fail?

It works fine on my machine.. :unamused:

**Comment by brson** [comment]
It's creating a path from the current working directory to somewhere in the source directory. This should be fine if building from the source directory, but if building out of tree (running configure from somewhere other than the source directory) it will produce the wrong path. Again just a guess.

**Comment by azerupi** [comment]
That makes sense. Unfortunately I don't see any variable that points to the root directory and I cant use a relative path from `src` or `tgt` either because they are not fixed and are given as arguments.

Is there something I could use that will always point to the right directory?

**Comment by brson** [comment]
@azerupi I'd suggest following [the example here](https://github.com/azerupi/rust/blob/doc-experiments/src/rustbook/build.rs#L163) and using `include_bytes!` to include these resources in the rustbook binary and then output them as needed at runtime.

**Comment by brson** [comment]
Sorry, didn't mean to close.

**Comment by Gankro** [comment]
@bors r-

Closing and Opening seemed to accidentally necro the build.

cc @barosl

**Comment by Gankro** [comment]
@azerupi what's your status?

**Comment by azerupi** [comment]
> what's your status?

Yes sorry for the wait, I have been distracted by other things (studies).
I will try to update the PR today with the needed modifications so it can finally be merged.

**Comment by azerupi** [comment]
I think it's good now, can someone review it?

**Comment by steveklabnik** [comment]
Looks good, let's see what bors says :)

**Comment by azerupi** [comment]
Great ! Finally, thanks to everyone for your patience :)
